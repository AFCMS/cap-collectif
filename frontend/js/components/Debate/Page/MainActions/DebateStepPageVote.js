// @flow
import React, { useState } from 'react';
import { useIntl, type IntlShape } from 'react-intl';
import { m as motion } from 'framer-motion';
import css from '@styled-system/css';
import Flex from '~ui/Primitives/Layout/Flex';
import Button from '~ds/Button/Button';
import Icon from '~ds/Icon/Icon';
import AddDebateVoteMutation from '~/mutations/AddDebateVoteMutation';
import LoginOverlay from '~/components/Utils/LoginOverlay';
import { mutationErrorToast } from '~/components/Utils/MutationErrorToast';
import { type VoteState } from './DebateStepPageVoteAndShare';
import type { AppBoxProps } from '~ui/Primitives/AppBox.type';

type Props = {|
  ...AppBoxProps,
  +debateId: string,
  +isAuthenticated: boolean,
  +onSuccess: VoteState => void,
  +viewerHasArgument: boolean,
|};

const voteForDebate = (
  debateId: string,
  type: 'FOR' | 'AGAINST',
  intl: IntlShape,
  onSuccess: VoteState => void,
  isAuthenticated: boolean,
  viewerHasArgument: boolean,
) => {
  return AddDebateVoteMutation.commit({ input: { debateId, type }, isAuthenticated })
    .then(response => {
      if (response.addDebateVote?.errorCode) {
        mutationErrorToast(intl);
      } else onSuccess(viewerHasArgument ? 'ARGUMENTED' : 'VOTED');
    })
    .catch(() => {
      mutationErrorToast(intl);
    });
};

const buttonColor = (styleColor: string, disabled: boolean) => ({
  opacity: disabled ? '.6' : '1',
  color: `${styleColor}.500`,
  backgroundColor: `${styleColor}.100`,
  borderColor: `${styleColor}.500`,
  border: '1px solid',
  '&:hover': {
    boxShadow: '0px 10px 50px 0px rgba(0, 0, 0, 0.15)',
  },
  '&:focus': {
    backgroundColor: `${styleColor}.500`,
    color: 'white',
    borderColor: `${styleColor}.500`,
  },
});

const Container = motion.custom(Flex);

export const DebateStepPageVote = ({
  debateId,
  isAuthenticated,
  onSuccess,
  viewerHasArgument,
  ...props
}: Props) => {
  const intl = useIntl();
  const [isHover, setIsHover] = useState<'FOR' | 'AGAINST' | false>(false);

  return (
    <Container
      transition={{ duration: 0.5 }}
      initial={{ opacity: 0 }}
      animate={{ opacity: 1 }}
      direction="row"
      alignItems="center"
      spacing={6}
      justifyContent="center"
      width="100%"
      {...props}>
      <LoginOverlay>
        <Button
          onMouseEnter={() => setIsHover('FOR')}
          onMouseLeave={() => setIsHover(false)}
          css={css(buttonColor('green', isHover === 'AGAINST'))}
          variantSize="big"
          aria-label={intl.formatMessage({ id: 'global.for' })}
          onClick={() =>
            voteForDebate(debateId, 'FOR', intl, onSuccess, isAuthenticated, viewerHasArgument)
          }>
          <Icon name="THUMB_UP" />
        </Button>
      </LoginOverlay>
      <LoginOverlay>
        <Button
          onMouseEnter={() => setIsHover('AGAINST')}
          onMouseLeave={() => setIsHover(false)}
          css={css(buttonColor('red', isHover === 'FOR'))}
          variantSize="big"
          aria-label={intl.formatMessage({ id: 'global.against' })}
          onClick={() =>
            voteForDebate(debateId, 'AGAINST', intl, onSuccess, isAuthenticated, viewerHasArgument)
          }>
          <Icon name="THUMB_DOWN" />
        </Button>
      </LoginOverlay>
    </Container>
  );
};

export default DebateStepPageVote;
