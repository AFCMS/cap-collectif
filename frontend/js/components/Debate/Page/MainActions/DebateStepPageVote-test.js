// @flow
/* eslint-env jest */
import * as React from 'react';
import { shallow } from 'enzyme';
import { DebateStepPageVote } from './DebateStepPageVote';

describe('<DebateStepPageVote/>', () => {
  it('renders correcty', () => {
    const wrapper = shallow(
      <DebateStepPageVote isAuthenticated debateId="debateId" onSuccess={jest.fn()} />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
