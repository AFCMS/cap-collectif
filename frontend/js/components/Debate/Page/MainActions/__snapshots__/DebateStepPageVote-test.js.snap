// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DebateStepPageVote/> should renders correctly 1`] = `
<div
  class="sc-bdnylx Flex___StyledAppBox-ll88x1-0 fBUPCS dzUdRG"
  style="opacity:0"
>
  <button
    class="sc-bdnylx Button__ButtonInner-sc-1rfi9ir-0 cTWwWU gQREUn DebateStepPageVote___StyledButton-sc-1t1mb78-0 dOSxvR"
    type="button"
  >
    <iconmock
      class="sc-bdnylx Icon__IconInner-sc-19k5t1s-0 Icon___StyledIconInner-sc-19k5t1s-1 ldGFSW jSXedj KA-deO icon"
    />
    global.for
  </button>
  <button
    class="sc-bdnylx Button__ButtonInner-sc-1rfi9ir-0 cTWwWU gQREUn DebateStepPageVote___StyledButton2-sc-1t1mb78-1 kZJOtr"
    type="button"
  >
    <iconmock
      class="sc-bdnylx Icon__IconInner-sc-19k5t1s-0 Icon___StyledIconInner-sc-19k5t1s-1 ldGFSW jSXedj KA-deO icon"
    />
    global.against
  </button>
</div>
`;

exports[`<DebateStepPageVote/> should renders correctly when not authenticated 1`] = `
<div
  class="sc-bdnylx Flex___StyledAppBox-ll88x1-0 fBUPCS dzUdRG"
  style="opacity:0"
>
  <button
    aria-controls="popover-overlay"
    aria-expanded="false"
    aria-haspopup="dialog"
    class="sc-bdnylx Button__ButtonInner-sc-1rfi9ir-0 cTWwWU gQREUn DebateStepPageVote___StyledButton-sc-1t1mb78-0 dOSxvR"
    type="button"
  >
    <iconmock
      class="sc-bdnylx Icon__IconInner-sc-19k5t1s-0 Icon___StyledIconInner-sc-19k5t1s-1 ldGFSW jSXedj KA-deO icon"
    />
    global.for
  </button>
  <div
    aria-label="vote.popover.title"
    class="sc-bdnylx LoginOverlay__PopoverContainer-c0lhnx-0 cNFNYl rLblB"
    data-dialog="true"
    hidden=""
    id="login-popover"
    role="dialog"
    style="display:none;position:fixed;left:100%;top:100%"
    tabindex="0"
  >
    <div
      class="sc-bdnylx LoginOverlay__Arrow-c0lhnx-2 kGqIZa hoKyqX"
      style="font-size:30px;width:1em;height:1em;pointer-events:none;bottom:100%"
    >
      <svg
        style="transform:rotateZ(360deg)"
        viewBox="0 0 30 30"
      >
        <path
          class="stroke"
          d="M23.7,27.1L17,19.9C16.5,19.3,15.8,19,15,19s-1.6,0.3-2.1,0.9l-6.6,7.2C5.3,28.1,3.4,29,2,29h26 C26.7,29,24.6,28.1,23.7,27.1z"
        />
        <path
          class="fill"
          d="M23,27.8c1.1,1.2,3.4,2.2,5,2.2h2H0h2c1.7,0,3.9-1,5-2.2l6.6-7.2c0.7-0.8,2-0.8,2.7,0L23,27.8L23,27.8z"
        />
      </svg>
    </div>
    <div
      class="sc-bdnylx eWkbGr"
    >
      <p
        class="sc-bdnylx YLtjJ"
      >
        vote.popover.title
      </p>
      <div
        class="sc-bdnylx Flex___StyledAppBox-ll88x1-0 kIsWfm cRYDhl"
      >
        <p
          class="sc-bdnylx jEhuac"
        >
          vote.popover.body
        </p>
        <span
          class="sc-bdnylx bgIGJQ"
          clip="rect(0px, 0px, 0px, 0px)"
        >
          <button
            class="sc-bdnylx Button__ButtonInner-sc-1rfi9ir-0 khpBZc gQREUn"
            type="button"
          >
            global.close
          </button>
        </span>
        <span>
          <button
            aria-label="open.connection_modal"
            class="center-block btn-block login-button btn btn-success"
            destination="http://localhost"
            type="button"
          >
            global.login
          </button>
        </span>
      </div>
    </div>
  </div>
  <button
    aria-controls="popover-overlay"
    aria-expanded="false"
    aria-haspopup="dialog"
    class="sc-bdnylx Button__ButtonInner-sc-1rfi9ir-0 cTWwWU gQREUn DebateStepPageVote___StyledButton2-sc-1t1mb78-1 kZJOtr"
    type="button"
  >
    <iconmock
      class="sc-bdnylx Icon__IconInner-sc-19k5t1s-0 Icon___StyledIconInner-sc-19k5t1s-1 ldGFSW jSXedj KA-deO icon"
    />
    global.against
  </button>
  <div
    aria-label="vote.popover.title"
    class="sc-bdnylx LoginOverlay__PopoverContainer-c0lhnx-0 cNFNYl rLblB"
    data-dialog="true"
    hidden=""
    id="login-popover"
    role="dialog"
    style="display:none;position:fixed;left:100%;top:100%"
    tabindex="0"
  >
    <div
      class="sc-bdnylx LoginOverlay__Arrow-c0lhnx-2 kGqIZa hoKyqX"
      style="font-size:30px;width:1em;height:1em;pointer-events:none;bottom:100%"
    >
      <svg
        style="transform:rotateZ(360deg)"
        viewBox="0 0 30 30"
      >
        <path
          class="stroke"
          d="M23.7,27.1L17,19.9C16.5,19.3,15.8,19,15,19s-1.6,0.3-2.1,0.9l-6.6,7.2C5.3,28.1,3.4,29,2,29h26 C26.7,29,24.6,28.1,23.7,27.1z"
        />
        <path
          class="fill"
          d="M23,27.8c1.1,1.2,3.4,2.2,5,2.2h2H0h2c1.7,0,3.9-1,5-2.2l6.6-7.2c0.7-0.8,2-0.8,2.7,0L23,27.8L23,27.8z"
        />
      </svg>
    </div>
    <div
      class="sc-bdnylx eWkbGr"
    >
      <p
        class="sc-bdnylx YLtjJ"
      >
        vote.popover.title
      </p>
      <div
        class="sc-bdnylx Flex___StyledAppBox-ll88x1-0 kIsWfm cRYDhl"
      >
        <p
          class="sc-bdnylx jEhuac"
        >
          vote.popover.body
        </p>
        <span
          class="sc-bdnylx bgIGJQ"
          clip="rect(0px, 0px, 0px, 0px)"
        >
          <button
            class="sc-bdnylx Button__ButtonInner-sc-1rfi9ir-0 khpBZc gQREUn"
            type="button"
          >
            global.close
          </button>
        </span>
        <span>
          <button
            aria-label="open.connection_modal"
            class="center-block btn-block login-button btn btn-success"
            destination="http://localhost"
            type="button"
          >
            global.login
          </button>
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`<DebateStepPageVote/> should renders correctly with anonymous participation 1`] = `
<div
  class="sc-bdnylx Flex___StyledAppBox-ll88x1-0 fBUPCS dzUdRG"
  style="opacity:0"
>
  <button
    class="sc-bdnylx Button__ButtonInner-sc-1rfi9ir-0 cTWwWU gQREUn DebateStepPageVote___StyledButton-sc-1t1mb78-0 dOSxvR"
    type="button"
  >
    <iconmock
      class="sc-bdnylx Icon__IconInner-sc-19k5t1s-0 Icon___StyledIconInner-sc-19k5t1s-1 ldGFSW jSXedj KA-deO icon"
    />
    global.for
  </button>
  <button
    class="sc-bdnylx Button__ButtonInner-sc-1rfi9ir-0 cTWwWU gQREUn DebateStepPageVote___StyledButton2-sc-1t1mb78-1 kZJOtr"
    type="button"
  >
    <iconmock
      class="sc-bdnylx Icon__IconInner-sc-19k5t1s-0 Icon___StyledIconInner-sc-19k5t1s-1 ldGFSW jSXedj KA-deO icon"
    />
    global.against
  </button>
</div>
`;

exports[`<DebateStepPageVote/> should renders correctly with email not confirmed argument 1`] = `
<div
  class="sc-bdnylx Flex___StyledAppBox-ll88x1-0 fBUPCS dzUdRG"
  style="opacity:0"
>
  <button
    class="sc-bdnylx Button__ButtonInner-sc-1rfi9ir-0 cTWwWU gQREUn DebateStepPageVote___StyledButton-sc-1t1mb78-0 dOSxvR"
    type="button"
  >
    <iconmock
      class="sc-bdnylx Icon__IconInner-sc-19k5t1s-0 Icon___StyledIconInner-sc-19k5t1s-1 ldGFSW jSXedj KA-deO icon"
    />
    global.for
  </button>
  <button
    class="sc-bdnylx Button__ButtonInner-sc-1rfi9ir-0 cTWwWU gQREUn DebateStepPageVote___StyledButton2-sc-1t1mb78-1 kZJOtr"
    type="button"
  >
    <iconmock
      class="sc-bdnylx Icon__IconInner-sc-19k5t1s-0 Icon___StyledIconInner-sc-19k5t1s-1 ldGFSW jSXedj KA-deO icon"
    />
    global.against
  </button>
</div>
`;
